core/src/main/java/org/elasticsearch/common/lucene/search/MatchNoDocsQuery.java

```diff
@@ -0,0 +1,46 @@
+package org.elasticsearch.common.lucene.search;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.*;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * A query that matches no documents and prints the reason why in the toString method.
+ */
+public class MatchNoDocsQuery extends Query {
+    /**
+     * The reason why the query does not match any document.
+     */
+    private final String reason;
+
+    public MatchNoDocsQuery(String reason) {
+        this.reason = reason;
+    }
+
+    public Weight createWeight(IndexSearcher searcher, boolean needsScores) throws IOException {
+        return new ConstantScoreWeight(this) {
+            @Override
+            public void extractTerms(Set<Term> terms) {
+            }
+
+            @Override
+            public Explanation explain(LeafReaderContext context, int doc) throws IOException {
+                return Explanation.noMatch(reason);
+            }
+
+            @Override
+            public Scorer scorer(LeafReaderContext context) throws IOException {
+                return null;
+            }
+        };
+    }
+
+    @Override
+    public String toString(String field) {
+        return "#false(\"" + reason + "\")";
```

Comment:
The use of `#` might make queries confusing since it is also used for filter clauses.